File: server/chat/message.go
package chat

import "fmt"

// Message stores information about a chat message
type Message struct {
	UserName  string `json:"userName"`
	Body      string `json:"body"`
	Action    string `json:"action"`
	Timestamp string `json:"timestamp"`
}

func (message *Message) String() string {
	return fmt.Sprintf("[%s] %s %s: %s",
		message.Timestamp, message.UserName, message.Action, message.Body)
}

File: client/src/app/chat/shared/message.ts
import { Action } from './action';

export class Message {
  userName?: string;
  body?: any;
  timestamp?: string;
  action?: Action;

  constructor(options?: {
    userName?: string,
    body?: any,
    timestamp?: string,
    action?: Action,
  }) {
    this.userName = options.userName || '';
    this.body = options.body;
    this.timestamp = options.timestamp;
    this.action = options.action;
  }

  get date(): Date {
    if (!this.timestamp) return null;
    return new Date(parseInt(this.timestamp));
  }
}
